src/main/java/com/commercetools/sync/taxcategories/utils/TaxRatesUpdateActionUtils.java

```diff
@@ -0,0 +1,234 @@
+     */
+    @Nonnull
+    private static List<UpdateAction<TaxCategory>> buildUpdateActions(
+        @Nonnull final List<TaxRate> oldTaxRates,
+        @Nonnull final List<TaxRateDraft> newTaxRatesDrafts)
+        throws BuildUpdateActionException {
+
+        try {
+            final List<UpdateAction<TaxCategory>> updateActions =
+                buildRemoveOrReplaceTaxRateUpdateActions(
+                    oldTaxRates,
+                    newTaxRatesDrafts
+                );
+
+            updateActions.addAll(
+                buildAddTaxRateUpdateActions(
+                    oldTaxRates,
+                    newTaxRatesDrafts
+                )
+            );
+
+            return updateActions;
+        } catch (final DuplicateCountryCodeException exception) {
+            throw new BuildUpdateActionException(exception);
+        }
+    }
+
+    /**
+     * Checks if there are any tax rates which are not existing in the {@code newTaxRatesDrafts}.
+     * If there are, then "remove" tax rate update actions are built.
+     * Otherwise, if the tax rate still exists in the new draft, then compare the tax rate
+     * fields (amount, country, state, etc..), and add the computed actions to the list of update actions.
+     *
+     * @param oldTaxRates       the list of old {@link TaxRate}s.
+     * @param newTaxRatesDrafts the list of new {@link TaxRateDraft}s.
+     * @return a list of tax rate update actions if there are tax rates that are not existing
+     *         in the new draft. If the tax rate still exists in the new draft, then compare the fields, and add
+     *         the computed actions to the list of update actions.
+     *         Otherwise, if the tax rates are identical, an empty list is returned.
+     * @throws DuplicateCountryCodeException in case there are tax rates drafts with duplicate country codes.
+     */
+    @Nonnull
+    private static List<UpdateAction<TaxCategory>> buildRemoveOrReplaceTaxRateUpdateActions(
+        @Nonnull final List<TaxRate> oldTaxRates,
+        @Nonnull final List<TaxRateDraft> newTaxRatesDrafts) {
+
+        /*
+        For TaxRates only unique field is country code. So we are using country code for matching.
+        Representation of CTP error,
+            {
+                "statusCode": 400,
+                "message": "A duplicate value '{\"country\":\"DE\"}' exists for field 'country'.",
+                "errors": [
+                    {
+                        "code": "DuplicateField",
+                        ....
+                ]
+            }
+        * */
+        final Map<CountryCode, TaxRateDraft> newTaxRatesDraftsCountryMap =
+            newTaxRatesDrafts
+                .stream().collect(
+                toMap(TaxRateDraft::getCountry, taxRateDraft -> taxRateDraft,
+                    (taxRateDraftA, taxRateDraftB) -> {
+                        throw new DuplicateCountryCodeException(
```

Comment:
Why this validation is done so late? It could be validated together with the tax category key validation, right?

Reply:
Why validation is done by country only and not by the combination of country and state? For example US has different tax rates for states. Also matching is below would need to take states into account.

Reply:
Good catch! you are right, I could write a batch validator to make it failing fast. I did not consider states, as I was not aware of this case. I will test this with platform and check what could be done here, those parts definitely need some changes.

Reply:
Move the checking to TaxCategorySync.java and do the checking by the combination of country and state.
https://github.com/commercetools/commercetools-sync-java/pull/524/commits/f5272b67147e1ea92baa9b56578f5e67d03c7c2e
https://github.com/commercetools/commercetools-sync-java/pull/524/commits/218e9cc16f42d4e8ea0b7a7cd786c2480f5920c3
https://github.com/commercetools/commercetools-sync-java/pull/524/commits/7f18e2c902f8ec1cb9bf580950f687200e885345